### ISSUE TITLE
Question about singleton bound

### ISSUE THREAD

**Issue by huangdijia** [laravel/octane#265]
- Octane Version: 0.5.0
- Laravel Version: 8.40.0
- PHP Version: 8.0.6
- Server & Version: Swoole 4.5.9

### Description:

- Register a singleton service

```php
use App\Service;

/**
 * Register any application services.
 *
 * @return void
 */
public function register()
{
    $this->app->singleton(Service::class, function ($app) {
        var_dump(__METHOD__);
        return new Service($app['request']);
    });
}
```

- Call

```php
// routes/web.php

Route::get('/', function() {
    app(App\Service::class); // Will output 'string(23) "App\Providers\{closure}"' every requests
});
```

### Steps To Reproduce:

Just request `http://127.0.0.1:8000` in the browser。

**Comment by mfn** [comment]
This is a weird one.

I found https://divinglaravel.com/laravel-octane-bootstrapping-the-application-and-handling-requests
> Singletons that are resolved while handling requests won't persist tho, they will get constructed on every request when running under Octane.

🤷‍♀️

It seems singletons not resolved in providers but during requests are only singletons for this particular request?

I.e. if a resolve this singleton multiple times _within_ a request, it only gets resolved once. But it still gets resolved for each and every request.

That's weird. I would have expected that classes registered as singletons would be only initialized once "per worker" in octane.

What am I missing?

Similar to OP, see this example

- `\App\Providers\MyService`
  ```php
  <?php
  
  namespace App\Providers;
  
  use Illuminate\Support\Facades\Log;
  use Illuminate\Support\ServiceProvider;
  
  class MyService extends ServiceProvider
  {
      /**
       * Register services.
       *
       * @return void
       */
      public function register()
      {
          Log::debug(__CLASS__ . ' register');
  
          $this->app->singleton(\App\MyService::class);
      }
  
      /**
       * Bootstrap services.
       *
       * @return void
       */
      public function boot()
      {
          Log::debug(__CLASS__ . ' boot');
      }
  }
  ```
- `\App\MyService`:
  ```php
  <?php
  
  namespace App;
  
  use Illuminate\Support\Facades\Log;
  
  class MyService
  {
      public function __construct()
      {
          Log::debug(__CLASS__ . ' __construct');
      }
  }
  ```
- add provider to `config/app.php`
- adjust `routes/web.php`:
  ```php
  Route::get('/', function () {
      app(MyService::class);
      app(MyService::class);
      return view('welcome');
  });
  ```
- start octane with `./artisan octane:start --workers=1`
  (I'm using RoadRunner in my case)
- open http://127.0.0.1:8000/
- see the logs
  ```
  [2021-05-14 21:12:52] local.DEBUG: App\Providers\MyService register
  [2021-05-14 21:12:52] local.DEBUG: App\Providers\MyService boot
  [2021-05-14 21:12:52] local.DEBUG: App\MyService __construct
  … pause… and re-request http://127.0.0.1:8000/
  
  [2021-05-14 21:12:57] local.DEBUG: App\MyService __construct
  ```

**Comment by mfn** [comment]
Ok, found ~the~ a solution, same reference again https://divinglaravel.com/laravel-octane-bootstrapping-the-application-and-handling-requests
> In case you want to resolve some singletons while the worker is starting and you don't explicitly resolve them in your service provider, Octane allow you to set an array of services to pre-warm for you.

Or from the config https://github.com/laravel/octane/blob/5b255108088e969c1584fe275f44b747a2a71d36/config/octane.php#L106-L119

Adding the class there makes it truly resolve only once (per worker).

---
Though I'm not sure yet how 3rd party packages could make use of this behaviour without asking the user to be added there.

**Comment by aarondfrancis** [comment]
This was also really confusing to me. The docs state that

> For example, the register and boot methods of your application's service providers will only be executed once when the request worker initially boots. On subsequent requests, the same application instance will be reused.

I put this in my service provider:

```php
$this->app->singleton('baz', function () {
    info('making');
    return 'bar';
});
```

and this in a controller 

```php
app('baz');
```

And every request generates a new "making" log entry.

Adding it to the `warm` array makes that go away. I'm not sure if that's intended behavior and the docs are misleading, or if it's a bug.

Thanks @mfn for reporting what you found!

**Comment by fpolli** [comment]
> This was also really confusing to me. The docs state that
> 
> > For example, the register and boot methods of your application's service providers will only be executed once when the request worker initially boots. On subsequent requests, the same application instance will be reused.
> 
> I put this in my service provider:
> 
> ```
> $this->app->singleton('baz', function () {
>     info('making');
>     return 'bar';
> });
> ```
> 
> and this in a controller
> 
> ```
> app('baz');
> ```
> 
> And every request generates a new "making" log entry.
> 
> Adding it to the `warm` array makes that go away. I'm not sure if that's intended behavior and the docs are misleading, or if it's a bug.
> 
> Thanks @mfn for reporting what you found!

My understanding of all this is that your singleton call BINDS the class but it does not RESOLVE it. That code simply tells Laravel how to resolve when the time comes. When you do nothing more, Laravel won't actually RESOLVE the binding until you use it in a request, and then it acts as above, where every request it gets remade and only gets reused if you use it again with the same request.

Putting it in the 'warm' array changes that behavior because Octane RESOLVES the bindings in the warm array on worker bootup, so the singleton is ready for requests.

What I have not been able to find ANYWHERE online is an explanation or example of HOW TO RESOLVE IN THE PROVIDER. The diving laravel post tells us that you can resolve the binding in the provider and have the singleton ready for all requests, but as far as I can tell, not even he tells us HOW TO F***ING DO IT. My guess is that you use $this->app->make() in the boot method, but I am not certain and I don't really know how to do it, because don't you have to assign that to a variable?

Very frustrating gap in documentation and third-party posts. Everybody tells us how to BIND a service, but explanation on the RESOLVING part is sparse and incomplete.

**Comment by rasmuscnielsen** [comment]
Damn. 4 years later and I just spent half a day figuring this out the hard way too...

**TLDR:**

- Instances **will** be reused across requests
- Singletons **might** be reused across requests (if resolved during boot)
- Scoped **will never** be reused across requests

_____

It's especially confusing in this part of the docs where it implies that singletons will be reused cross requests, while `scoped` won't. 

<img width="666" height="1139" alt="Image" src="https://github.com/user-attachments/assets/c19534f8-da2e-4340-b822-b2073df8723a" />

While not technically wrong, it does give a false expectation of how singletons would behave.  

But @fpolli is spot on. From my Laravel 12 tests:

In AppServiceProvider@register

```php 
$this->app->singleton(Counter::class);

// Result: Not reused across requests  ❌
```


```php 
$this->app->singleton(Counter::class);
$this->app->make(Counter::class); // resolving the instance during registration or boot (!)

// Result: Reused across requests ✅
```


```php 
$this->app->instance(Counter::class, new Counter);

// Result: Reused across requests ✅
```


```php 
$this->app->scoped(Counter::class);

// Result: Not reused across requests ❌ (as intended and described in docs)
```

```php 
$this->app->scoped(Counter::class);
$this->app->make(Counter::class); // resolving the instance during registration or boot

// Result: Not reused across requests ❌ (as intended and described in docs) - key differentiator from singleton! 
```
